[InitUndef] Enable the InitUndef pass on non-AMDGPU targets

nikic opened the pull request:
The InitUndef pass works around a register allocation issue, where undef operands can be allocated to the same register as early-clobber result operands. This may lead to ISA constraint violations, where certain input and output registers are not allowed to overlap.

Originally this pass was implemented for RISCV, and then extended to ARM in #77770. I've since removed the target-specific parts of the pass in #106744 and #107885. This PR reduces the pass to use a single requiresDisjointEarlyClobberAndUndef() target hook and enables it by default. The hook is disabled for AMDGPU, because overlapping early-clobber and undef operands are known to be safe for that target, and we get significant codegen diffs otherwise.

The motivating case is the one in arm64-ldxr-stxr.ll, where we were previously incorrectly allocating a stxp input and output to the same register.

nikic committed:
[InitUndef] Enable the InitUndef pass on all targets

The InitUndef pass works around a register allocation issue, where
undef operands can be allocated to the same register as early-clobber
result operands. This may lead to ISA constraint violations, where
certain input and output registers are not allowed to overlap.

Originally this pass was implemented for RISCV, and then extended
to ARM in #77770. I've since removed the target-specific parts of
the pass in #106744 and #107885. This PR now enables the pass for
all targets.

The motivating case is the one in arm64-ldxr-stxr.ll for the AArch64
target, where we were previously incorrectly allocating a stxp input
and output to the same register.

jayfoad commented:
I don't think this has any benefit for AMDGPU. We use early-clobber to avoid some outputs overlapping with defined inputs, but it's OK for them to overlap with undef inputs. One specific example:
```
v_mad_u64_u32 v[9:10], null, v2, v3, v[8:9]
```
Here we generated a 64-bit integer multiply-add (v[9:10] = zext(v2) * zext(v3) + v[8:9]) but we're only using the low part of the result (v9 = v2 * v3 + v8). This means that the v9 (the high part of the addend input v[8:9]) is an undef input so it's OK for it to overlap with the output, even though the output is early-clobber.

jayfoad commented:
The AMDGPU usage is a very literal interpretation of what "early clobber" means, i.e. (some part of) the output might get written before (some parts of) the inputs are read.

For the architectural constraints in ARM and RISCV, is the output forbidden from overlapping with any input, or with one specific input? If the latter, maybe a better way to model it would be with something like `Constraints = "$dst != $src2"` in the tablegen definition of the instruction -- but that would require support in all register allocators, and I don't know how much effort anyone wants to put into better handling for this minor codegen issue.

Stylie777 commented:
@jayfoad I can only speak for Arm and AArch64 but the implementation we use to not reuse registers, unlike what is done in AMDGPU as you described. 

I think it would be easier here to just ensure AMDGPU is disabled, by keeping the `SupportsInitUndef` function and adding an implementation for AArch64. There does not seem to be any other architectures affected by this change.

nikic commented:
If this is undesirable for AMDGPU, I can restore the target hook. A caveat is that I think this is always required for inline asm (because we have no knowledge over whether it's safe or not in that case), but if needed we could explicitly check for that case separately from the rest.

nikic committed:
Restore hook, only enable on aarch64

BeMg commented:
> The AMDGPU usage is a very literal interpretation of what "early clobber" means, i.e. (some part of) the output might get written before (some parts of) the inputs are read.
> 
> For the architectural constraints in ARM and RISCV, is the output forbidden from overlapping with any input, or with one specific input? If the latter, maybe a better way to model it would be with something like `Constraints = "$dst != $src2"` in the tablegen definition of the instruction -- but that would require support in all register allocators, and I don't know how much effort anyone wants to put into better handling for this minor codegen issue.

In RISC-V, we are trying to resolve a case involving the `vrgather.vv` instruction by InitUndef, where the destination register cannot overlap with any source register; otherwise, it is considered an illegal instruction.

nikic commented:
I've updated the PR to restore the hook and only enable it on AArch64 additionally. I've renamed the hook to make it clearer when you are supposed to enable it.

Though I'm wondering if it wouldn't be better to enable it by default and only disable on AMDGPU, as that is the conservatively correct behavior, and it doesn't seem like this affects register allocation/scheduling on other targets anyway.

jayfoad commented:
> > The AMDGPU usage is a very literal interpretation of what "early clobber" means, i.e. (some part of) the output might get written before (some parts of) the inputs are read.
> > For the architectural constraints in ARM and RISCV, is the output forbidden from overlapping with any input, or with one specific input? If the latter, maybe a better way to model it would be with something like `Constraints = "$dst != $src2"` in the tablegen definition of the instruction -- but that would require support in all register allocators, and I don't know how much effort anyone wants to put into better handling for this minor codegen issue.
> 
> In RISC-V, we are trying to resolve a case involving the `vrgather.vv` instruction by InitUndef, where the destination register cannot overlap with any source register; otherwise, it is considered an illegal instruction.

Thanks. But I also see cases like this:
```
lib/Target/RISCV/RISCVInstrInfoV.td
1114-// Set earlyclobber for following instructions for second and mask operands.
1115:// This has the downside that the earlyclobber constraint is too coarse and
1116-// will impose unnecessary restrictions by not allowing the destination to
1117-// overlap with the first (wide) operand.
```
which sound like they could benefit from a `!=` constraint.

arsenm commented on llvm/test/CodeGen/AArch64/arm64-ldxr-stxr.ll:
```diff
@@ -354,11 +354,10 @@ define dso_local i32 @test_store_release_i64(i32, i64 %val, ptr %addr) {
 ; CHECK-LABEL: test_stxp_undef:
 ; CHECK:       // %bb.0:
-; CHECK-NEXT:    stxp w8, x8, x1, [x0]
+; CHECK-NEXT:    stxp w8, x9, x1, [x0]
```
This could use an additional MIR test for something so specific

nikic replied:
The pass itself already has MIR tests. I don't think adding another MIR test for this case would be useful, because we care about the end-to-end behavior (from the interaction of three different passes) here.

What would be quite nice is if the old code would emit an error, as it's hard to spot the error otherwise. Currently it only generates an error when parsed back as inline assembly. Not sure whether there is a good way to share that validation, I'm not familiar with this part of the backend...

efriedma-quic replied:
AArch64InstrInfo::verifyInstruction is the place for target-specific verification of an instruction... but currently it only gets called with -verify-machineinstrs.  (It looks like AMDGPUAsmPrinter::emitInstruction calls verifyInstruction, but other targets currently don't do that.)

arsenm replied:
I only see 3 places using run-pass=init-undef, all in RISCV. You could just dump the MIR before init-undef, extract the one instruction, and make sure the pass isn't a no-op on AArch64

nikic replied:
Okay, if all we want to test is "the pass does something", then a MIR test makes sense. I've added one.

arsenm commented:
> Though I'm wondering if it wouldn't be better to enable it by default and only disable on AMDGPU, as that is the conservatively correct behavior, and it doesn't seem like this affects register allocation/scheduling on other targets anyway.

Yes, the default should always be conservatively correct. This type of stuff is very undiscoverable

nikic committed:
Enable by default, disable on AMDGPU

nikic committed:
Add MIR test

nikic commented:
> > Though I'm wondering if it wouldn't be better to enable it by default and only disable on AMDGPU, as that is the conservatively correct behavior, and it doesn't seem like this affects register allocation/scheduling on other targets anyway.
> 
> Yes, the default should always be conservatively correct. This type of stuff is very undiscoverable

Done.

arsenm commented on llvm/test/CodeGen/AArch64/init-undef.mir:
```diff
@@ -0,0 +1,44 @@
+# RUN: llc -mtriple=aarch64-- -run-pass=init-undef -o - %s | FileCheck %s
+
+--- |
+  define dso_local i32 @test_stxp_undef(ptr %p, i64 %x) #0 {
```
Don't need the IR section

arsenm commented on llvm/test/CodeGen/AArch64/init-undef.mir:
```diff
@@ -0,0 +1,44 @@
+name:            test_stxp_undef
+alignment:       4
+tracksRegLiveness: true
+registers:
+  - { id: 0, class: gpr64common }
+  - { id: 1, class: gpr64 }
+  - { id: 2, class: gpr32 }
+  - { id: 3, class: gpr64 }
```
```suggestion
```

arsenm commented on llvm/test/CodeGen/AArch64/init-undef.mir:
```diff
@@ -0,0 +1,44 @@
+  - { reg: '$x0', virtual-reg: '%0' }
+  - { reg: '$x1', virtual-reg: '%1' }
+body:             |
+  bb.0 (%ir-block.0):
```
```suggestion
  bb.0:
```

arsenm commented on llvm/test/CodeGen/AArch64/init-undef.mir:
```diff
@@ -0,0 +1,44 @@
+    %1:gpr64 = COPY $x1
+    %0:gpr64common = COPY $x0
+    %3:gpr64 = IMPLICIT_DEF
+    early-clobber %2:gpr32 = STXPX killed %3, %1, %0 :: (volatile store (s128) into %ir.p)
```
```suggestion
    early-clobber %2:gpr32 = STXPX killed %3, %1, %0 :: (volatile store (s128))
```

nikic committed:
Simplify MIR test
